### PR TITLE
refact(upgrade): added functionality to create and update the UpgradeTask CR

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**What this PR does / why we need it**:
This PR adds the code create or read from the UpgradeTask CR.
If the CR is not present job will create one for the resource. 
It also updates the CR with the current status of the upgrade.
This PR also includes the ability to upgrade multiple resources in one single job.

Sample job yaml will be
```
apiVersion: batch/v1
kind: Job
metadata:
  # VERIFY that you have provided a unique name for this upgrade job.
  # The name can be any valid K8s string for name. This example uses
  # the following convention: jiva-vol-<flattened-from-to-versions>-<pv-name>
  name: jiva-vol-100110-pvc-7904a228-ca69-11e9-9c4f-54e1ad5e8320
  
  # VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs

spec:
  backoffLimit: 4
  template:
    spec:
      
      # VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`
      serviceAccountName: openebs-maya-operator
      
      containers:
      - name:  upgrade
        args: 
        - "jiva-volume"
        - "--from-version=1.0.0"
        - "--to-version=1.1.0"
        
        # VERIFY that you have provided the correct cStor PV Name
        - "--pv-name=pvc-7904a228-ca69-11e9-9c4f-54e1ad5e8320"
        
        # Following are optional parameters
        # Log Level
        - "--v=4"
        
        # DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        tty: true 
        image: openebs/m-upgrade:ci
        imagePullPolicy: IfNotPresent
      restartPolicy: OnFailure
```

Sample Success UpgradeTask CR
```
apiVersion: openebs.io/v1alpha1
kind: UpgradeTask
metadata:
  creationTimestamp: "2019-08-29T15:17:33Z"
  generation: 9
  name: upgrade-jiva-volume-pvc-7904a228-ca69-11e9-9c4f-54e1ad5e8320
  namespace: openebs
  labels:
    app: app-name
  resourceVersion: "9464"
  selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/upgradetasks/upgrade-jiva-volume-pvc-7904a228-ca69-11e9-9c4f-54e1ad5e8320
  uid: 1ff2551d-ca70-11e9-9c4f-54e1ad5e8320
spec:
  fromVersion: 1.0.0
  imagePrefix: quay.io/openebs/
  imageTag: 1.1.0
  jivaVolume:
    pvName: pvc-7904a228-ca69-11e9-9c4f-54e1ad5e8320
  toVersion: 1.1.0
status:
  completedTime: "2019-08-29T15:18:27Z"
  phase: Success
  retries: 0
  startTime: "2019-08-29T15:17:33Z"
  upgradeDetailedStatuses:
  - lastUpdatedAt: "2019-08-29T15:17:33Z"
    message: Pre-upgrade steps were successful
    phase: Completed
    startTime: "2019-08-29T15:17:33Z"
    step: PRE_UPGRADE
  - lastUpdatedAt: "2019-08-29T15:17:59Z"
    message: Replica upgrade was successful
    phase: Completed
    startTime: "2019-08-29T15:17:33Z"
    step: REPLICA_UPGRADE
  - lastUpdatedAt: "2019-08-29T15:18:26Z"
    message: Target upgrade was successful
    phase: Completed
    startTime: "2019-08-29T15:17:59Z"
    step: TARGET_UPGRADE
```
Sample Error UpgradeTask CR
```
apiVersion: openebs.io/v1alpha1
kind: UpgradeTask
metadata:
  creationTimestamp: "2019-08-29T15:24:13Z"
  generation: 20
  name: upgrade-jiva-volume-pvc-792c8504-ca69-11e9-9c4f-54e1ad5e8320
  namespace: openebs
  labels:
    app: app-name
  resourceVersion: "10682"
  selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/upgradetasks/upgrade-jiva-volume-pvc-792c8504-ca69-11e9-9c4f-54e1ad5e8320
  uid: 0e7233c3-ca71-11e9-9c4f-54e1ad5e8320
spec:
  fromVersion: 1.0.0
  imagePrefix: quay.io/openebs/
  imageTag: 1.1.0
  jivaVolume:
    pvName: pvc-792c8504-ca69-11e9-9c4f-54e1ad5e8320
  toVersion: 1.1.0
status:
  completedTime: "2019-08-29T15:25:55Z"
  phase: Error
  retries: 5
  startTime: "2019-08-29T15:24:13Z"
  upgradeDetailedStatuses:
  - lastUpdatedAt: "2019-08-29T15:25:55Z"
    message: failed to get target details
    phase: Errored
    reason: failed to get controller deployment
    startTime: "2019-08-29T15:25:55Z"
    step: PRE_UPGRADE
```
If the CRs is already created then the CR label can be passed instead of args in the job to upgrade multiple resources of similar kind in one job.

Sample UpgradeTask job
```
apiVersion: batch/v1
kind: Job
metadata:
  # VERIFY that you have provided a unique name for this upgrade job.
  # The name can be any valid K8s string for name. This example uses
  # the following convention: jiva-vol-<flattened-from-to-versions>-<pv-name>
  name: jiva-volume-100110-pvc-794c53a7-ca69-11e9-9c4f-54e1ad5e8320
  
  # VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs

spec:
  template:
    spec:
      
      # VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`
      serviceAccountName: openebs-maya-operator 
      containers:
      - name:  upgrade
        args:
        - "resource"
        
        # Following are optional parameters
        # Log Level
        - "--v=4"
        
        env:
        - name: OPENEBS_NAMESPACE
          value: "openebs"
        - name: UPGRADE_TASK_LABEL
          value: "app=app-name"
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        tty: true 
        image: openebs/m-upgrade:ci 
        imagePullPolicy: IfNotPresent
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go [hunk 1]
```diff
@@ -275,52 +280,352 @@ func getPVCDeploymentsNamespace(
...
+		if !quorum && syncedReplicas > (replicationFactor/2) {
+			glog.Infof("Synced replica quorum is reached")
+			quorum = true
+		}
+	}
+	glog.Infof("Replica syncing complete")
+	return nil
+}
+
+func jivaUpgrade(pvName, openebsNamespace string) (*utask.UpgradeTask, error) {
```
Comment (kmova) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go]: Can the upgrade functions like this be refactored to reduce the number of lines to less than 100?

File: cmd/upgrade/executor/resource.go [hunk 2]
```diff
@@ -52,28 +53,41 @@ func NewUpgradeResourceJob() *cobra.Command {
...
+			upgradeTaskCRName := getUpgradeTaskLabel()
+			openebsNamespace := getOpenEBSNamespace()
+			upgradeTaskList, err := utask.NewKubeClient().
+				WithNamespace(openebsNamespace).
+				List(metav1.ListOptions{
+					LabelSelector: upgradeTaskCRName,
+				})
+			util.CheckErr(err, util.Fatal)
+
+			for _, cr := range upgradeTaskList.Items {
```
Comment (kmova) [inline cmd/upgrade/executor/resource.go]: We should probably add a check here to make sure atleast one resource is found to upgrade. Can be taken up as next PR.

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go [hunk 3]
```diff
@@ -275,52 +280,281 @@ func getPVCDeploymentsNamespace(
...
+		return options.utaskObj, err
+	}
+
+	// TargetUpgrade
+	err = options.targetUpgrade(pvName, openebsNamespace)
+	if err != nil {
+		return options.utaskObj, err
+	}
+
+	// Verify
```
Comment (kmova) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go]: It is possible that on some setups pod/exec will be disabled. In this case, we should patch the status with a warning message and provide information on how to verify. Alternatively, we can probably use the m-apiserver api to list the jiva volume and verify the status. Also, can be taken in a different PR.

File: cmd/upgrade/executor/resource.go (hunk 2 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/resource.go]: Done

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go (hunk 3 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go]: Added a check for the cases in which `exec` into the pods is `Forbidden`. This required removing wrapper error from some functions as wrapping the error was removing the reason of the error.

File: pkg/kubernetes/pod/v1alpha1/kubernetes.go [hunk 4]
```diff
@@ -89,11 +89,7 @@ func defaultExec(
 	// for transporting shell-style streams
 	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
 	if err != nil {
-		return nil, errors.Wrapf(
-			err,
-			"failed to exec into pod {%s}: failed to connect to the provided server",
-			name,
-		)
+		return nil, err
```
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply pkg/kubernetes/pod/v1alpha1/kubernetes.go]: Some BDD use this functions but the changes won't impact them as this wrapper message is the first wrapper and the actual k8s error also include the name of the pod and the error stack trace will still contain the entire flow of the caller.
